### PR TITLE
Import I-Hong's improvments

### PR DESCRIPTION
**Pull request by bdwyer2**
Imported I-Hong's improvements to the Facial Recognizer model. This allows the model to run in real time.

All the credit for the code in this PR goes to @ihjhuo

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,22 @@
-# IBM Code Model Asset Exchange: Facial Age Estimator
+# IBM Code Model Asset Exchange:Age Estimation with SSR-Net
 
-This repository contains code to instantiate and deploy a facial age estimation model. The model detects faces in an image, extracts facial features for each face detected and finally predicts the age of each face. The model uses a coarse-to-fine strategy to perform multi-class classification and regression for age estimation. The input to the model is an image and the output is a list of estimated ages and bounding box coordinates of each face detected in the image. The format of the bounding box coordinates is `[xmin, ymin, width, height]`.
+The SSR-Net gives Age Estimation from giving an image or video. The model first detects human face, then automatically generate feature vector to represent every detected human face.  The model takes a coarse-to-fine strategy to perform multi-class classification and regression for age estimatation. The output of the model provides gender, age, and also bounding boxes of a given image or video.
```
Comment (frreiss) [inline README.md]: Jargon. Bring back the old, more clear verbiage here.

File: api/model.py [hunk 2]
```diff
@@ -27,14 +27,14 @@ def get(self):
 
 # Creating a JSON response model: https://flask-restplus.readthedocs.io/en/stable/marshalling.html#the-api-model-factory
 label_prediction = api.model('LabelPrediction', {
-    'age_estimation': fields.Integer(required=True, description='Estimated age for the face'),
-    'face_box': fields.List(fields.Integer(required=True), description='Bounding box coordinates for the face')
+    'age_estimation': fields.List(fields.Float(required=True)),
```
Comment (frreiss) [inline api/model.py]: Please add the description back in.

File: api/model.py [hunk 3]
```diff
@@ -27,14 +27,14 @@ def get(self):
...
+    'face_box': fields.List(fields.Float(required=True))
```
Comment (frreiss) [inline api/model.py]: Here too.

File: api/model.py [hunk 4]
```diff
@@ -27,14 +27,14 @@ def get(self):
...
 })
 
 
 predict_response = api.model('ModelPredictResponse', {
     'status': fields.String(required=True, description='Response status message'),
-    'predictions': fields.List(fields.Nested(label_prediction), description='Predicted age and bounding box for each detected face')
+    'predictions': fields.List(fields.Nested(label_prediction), description='Age estimation and face bounding box')
```
Comment (frreiss) [inline api/model.py]: Description should make it clear that this model detects multiple faces and returns a list of bounding boxes.

File: api/model.py [hunk 5]
```diff
@@ -66,5 +66,6 @@ def post(self):
             label_preds.append({'age_estimation':res[0]['age'],'face_box':res[0]['box']})
         result['predictions'] = label_preds
         result['status'] = 'ok'
+        print(label_preds)
```
Comment (frreiss) [inline api/model.py]: This debug output should go through a logging system, not get sent directly to STDOUT.

File: config.py [hunk 6]
```diff
@@ -12,14 +12,16 @@
...
 
 # default model
-MODELNAME = 'ssrnet_3_3_3_64_1.0_1.0.h5'
-DEFAULT_MODEL_PATH = 'assets/{}'.format(MODELNAME)
-MODEL_LICENSE = 'MIT'
+MODEL_NAME = 'ssrnet_3_3_3_64_1.0_1.0.h5'
+FACE_NAME = 'lbpcascade_frontalface_improved.xml'
+DEFAULT_MODEL_PATH = 'assets/{}'.format(MODEL_NAME)
+DEFAULT_DETECTOR_PATH = 'assets/{}'.format(FACE_NAME)
+MODEL_LICENSE = 'MODEL LICENSE'
```
Comment (frreiss) [inline config.py]: MIT, not MODEL LICENSE

File: config.py [hunk 7]
```diff
@@ -12,14 +12,16 @@
...
 
 MODEL_META_DATA = {
-    'id': 'ssrnet',
-    'name': 'SSR-Net Facial Age Estimator Model',
-    'description': 'SSR-Net Facial Recognition and Age Prediction model; trained using Keras on the IMDB-WIKI dataset',
-    'type': 'facial-recognition',
+    'id': '{} MODEL ID'.format(MODEL_NAME.lower()),
```
Comment (frreiss) [inline config.py]: This ID should be a string that is unique within MAX and has some connection to the MAX model name.

File: config.py [hunk 8]
```diff
@@ -12,14 +12,16 @@
...
+    'name': '{} MODEL NAME'.format(MODEL_NAME),
```
Comment (frreiss) [inline config.py]: This should be the same as the model name in the MAX site.

File: config.py [hunk 9]
```diff
@@ -12,14 +12,16 @@
...
+    'description': '{} MODEL DESCRIPTION'.format(MODEL_NAME),
```
Comment (frreiss) [inline config.py]: This should be a shorter version of the description in the README.md file.

File: core/backend.py [hunk 10]
```diff
@@ -48,14 +49,17 @@ def __init__(self, path=DEFAULT_MODEL_PATH):
...
+        self.detector_path = detector_path
+        self.model.load_weights(self.model_path)
 
         self.graph = tf.get_default_graph()
 
         logger.info('Loaded model')
 
     def predict(self, x):
         input_img = x
+        face_cascade = cv2.CascadeClassifier(self.detector_path)
```
Comment (frreiss) [inline core/backend.py]: This loading should happen once, not once per prediction.

File: core/backend.py [hunk 11]
```diff
@@ -99,7 +106,8 @@ def predict(self, x):
         # prediction results with BBX & AGES
         pred_res = []
         for i, d in enumerate(detected):
-            if d['confidence'] > 0.8:
-                pre_age=predicted_ages[i].astype(int)
-                pred_res.append([{'box': d['box'], 'age':pre_age}])
-        return pred_res
\ No newline at end of file
+            # if d['confidence'] > 0.95:
```
Comment (frreiss) [inline core/backend.py]: Why are we ignoring the returned confidence value? Shouldn't this be part of the information returned for each bounding box? Is this confidence on the bounding box location, or confidence of the age?

Comment (ihjhuo) [reply core/backend.py]: Using cv.CascadeClassifier only returns the bounding box

**Comment by bdwyer2** [comment]
@frreiss I addressed your comments. Is this okay to merge?

**Comment by frreiss** [comment]
@bdwyer2 Yes, please.

**Pull request merged**